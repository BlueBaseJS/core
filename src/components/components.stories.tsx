import * as React from 'react';
// tslint:disable-next-line:no-implicit-dependencies
import { storiesOf } from '@storybook/react';
import BR from '../index';
import ComponentStateButton from '../components/ComponentState/ComponentStateButton';

const stories = storiesOf('Avatar', module);

stories.add('ComponentStateButton', () => {
	return (
    <ComponentStateButton/>);
});
