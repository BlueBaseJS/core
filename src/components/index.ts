export * from './BlueBaseApp';
export * from './BlueBaseContent';
export * from './BlueBaseFilter';
export * from './ComponentState';
export * from './DataObserver';
export * from './DynamicIcon';
export * from './EmptyState';
export * from './ErrorObserver';
export * from './ErrorState';
export * from './Header';
export * from './HeaderBackButton';
export * from './HeaderTitle';
export * from './HomeScreen';
export * from './HoverObserver';
export * from './JsonSchema';
export * from './Link';
export * from './LoadingState';
export * from './Navigation';
export * from './NavigationActions';
export * from './Noop';
export * from './PluginIcon';
export * from './ReactLoadableLoading';
export * from './Redirect';
export * from './StatefulComponent';
export * from './WaitObserver';
