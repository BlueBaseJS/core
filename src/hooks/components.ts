import { BlueRain, BootOptions } from '../BlueRain';
import { HookInput } from '../registries';
import {
	BlueRainHook,
	ComponentState,
	DynamicIcon,
	EmptyState,
	ErrorState,
	JsonSchema,
	LoadingState,
	Noop,
	PluginIcon,
	SystemApp,
	SystemContent,
} from '../components';

import { ActivityIndicator, Button, Image, Text, View } from '../native';

export const components: { [key: string]: HookInput[] } = {

	'bluerain.components.register.internal': [{
		handler: async (bootOptions: BootOptions, _ctx: {}, BR: BlueRain) => {

			// await BR.Components.register('DynamicIcon', import('../components/DynamicIcon'));
			await BR.Components.register('BlueRainHook', BlueRainHook);
			await BR.Components.register('ComponentState', ComponentState);
			await BR.Components.register('DynamicIcon', DynamicIcon);
			await BR.Components.register('EmptyState', EmptyState);
			await BR.Components.register('ErrorState', ErrorState);
			await BR.Components.register('JsonSchema', JsonSchema);
			await BR.Components.register('LoadingState', LoadingState);
			await BR.Components.register('Noop', Noop);
			await BR.Components.register('PluginIcon', PluginIcon);
			await BR.Components.register('SystemApp', SystemApp);
			await BR.Components.register('SystemContent', SystemContent);
			await BR.Components.register('SystemFooter', Noop);
			await BR.Components.register('SystemHeader', Noop);

			await BR.Components.register('ActivityIndicator', ActivityIndicator);
			await BR.Components.register('Button', Button);
			await BR.Components.register('Image', Image);
			await BR.Components.register('Text', Text);
			await BR.Components.register('View', View);

			return bootOptions;
		},
		name: 'bluerain-components-register-internal-default',
		priority: 5,
	}]
};
