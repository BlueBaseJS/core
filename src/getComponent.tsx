import React from 'react';

import { useBlueBase } from './hooks';

/**
 * Resolves a component from BlueBase context, and returns it.
 * It is possible to pass multiple keys as backup. So if one component
 * is not found, the function will fallback and try to find next component.
 *
 * @param keys
 */
export function getComponent<T = any>(
	...keys: Array<string | React.ComponentType<any>>
): React.ComponentType<T> {
	if (keys.length === 0) {
		throw Error('getComponent method needs at least one key');
	}

	const displayName = keys.join('_');

	const BlueBaseComponent = (props: T) => {
		const BB = useBlueBase();

		// If there is no BlueBase context, throw an Error
		if (!BB) {
			throw Error(
				// eslint-disable-next-line max-len
				`Could not resolve component "${displayName}" in "getComponent" command. Reason: BlueBase context not found.`
			);
		}

		return React.createElement(BB.Components.resolveFromCache<T>(...(keys as any)) as any, props as any);
	};

	BlueBaseComponent.displayName = displayName;
	return BlueBaseComponent;
}
