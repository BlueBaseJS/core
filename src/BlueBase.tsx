import { Analytics, Logger } from './api';
import {
	ComponentCollection,
	ComponentRegistry,
	ConfigCollection,
	ConfigRegistry,
	HookNestedCollection,
	HookRegistry,
	PluginCollection,
	PluginRegistry,
	ThemeCollection,
	ThemeRegistry,
} from './registries';
import { BlueBaseProvider } from './Context';
import React from 'react';
import { ThemeProvider } from './themes';
import { getComponent } from './getComponent';
import systemHooks from './hooks';

export interface BootOptions {
	components: ComponentCollection,
	configs: ConfigCollection,
	hooks: HookNestedCollection,
	plugins: PluginCollection,
	themes: ThemeCollection,
	// routes: Plugin[]
}

export class BlueBase {

	// APIs
	public Analytics = new Analytics(this);
	public Logger = new Logger(this);

	// Registries
	public Components = new ComponentRegistry(this);
	public Configs = new ConfigRegistry(this);
	public Hooks = new HookRegistry(this);
	public Plugins = new PluginRegistry(this);
	public Themes = new ThemeRegistry(this);

	// Flags
	public booted = false;

	private bootOptions: BootOptions = {
		components: {},
		configs: {},
		hooks: {},
		plugins: [],
		themes: [],
	};

	public async boot(options?: Partial<BootOptions> & { children?: React.ReactNode }) {

		this.bootOptions = { ...this.bootOptions, ...options };

		await this.Hooks.registerNestedCollection(systemHooks);
		await this.Hooks.registerNestedCollection(this.bootOptions.hooks);

		await this.Hooks.run('bluebase.boot', this.bootOptions);

		const SystemApp = getComponent('SystemApp');

		// Set View
		// const SystemApp = this.Components.resolve('SystemApp');
		// SystemApp = await this.Hooks.run('bluebase.system.app', SystemApp);

		const BlueBaseRoot = () => (
			<BlueBaseProvider value={this}>
				<ThemeProvider>
					<SystemApp>
						{options && options.children}
					</SystemApp>
				</ThemeProvider>
			</BlueBaseProvider>
		);

		this.booted = true;

		return BlueBaseRoot;
	}
}