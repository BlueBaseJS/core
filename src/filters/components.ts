// tslint:disable: object-literal-sort-keys
import {
	ActivityIndicator,
	Button,
	Image,
	ImageBackground,
	Picker,
	Slider,
	Switch,
	Text,
	TextInput,
	View,
} from '../native';
import { BlueBase, BootOptions } from '../BlueBase';
import {
	BlueBaseContent,
	BlueBaseFilter,
	BlueBaseImage,
	ComponentState,
	DataObserver,
	DynamicIcon,
	EmptyState,
	ErrorObserver,
	ErrorState,
	FormattedMessage,
	Header,
	HeaderBackButton,
	HeaderTitle,
	HomeScreen,
	HoverObserver,
	Link,
	LoadingState,
	Navigation,
	NavigationActions,
	Noop,
	PluginIcon,
	Redirect,
	StatefulComponent,
	TouchableItem,
	WaitObserver,
} from '../components/';
import { FilterNestedCollection } from '../registries';
import { ThemeValue } from '../themes';

export const components: FilterNestedCollection = {
	/**
	 * Registers all components that ship with BlueBase
	 */
	'bluebase.components.register.internal': [
		{
			key: 'bluebase-components-register-internal-default',
			priority: 5,

			value: async (bootOptions: BootOptions, _ctx: {}, BB: BlueBase) => {
				await BB.Components.registerCollection({
					// BlueBase Components
					BlueBaseContent,
					BlueBaseFilter,
					BlueBaseImage,
					ComponentState,
					DataObserver,
					DynamicIcon,
					EmptyState,
					ErrorObserver,
					ErrorState,
					FormattedMessage,
					Header,
					HeaderBackButton,
					HeaderTitle,
					HomeScreen,
					HoverObserver,
					Icon: Noop,
					Link,
					LoadingState,
					Navigation,
					NavigationActions,
					Noop,
					PluginIcon,
					Redirect,
					StatefulComponent,
					TouchableItem,
					WaitObserver,

					// Native
					ActivityIndicator,
					Button,
					Image,
					ImageBackground,
					Picker,
					Slider,
					Switch,
					Text,
					TextInput,
					View,

					// Typography
					H1: {
						value: Text,
						styles: (theme: ThemeValue) => ({ root: theme.typography.h1 }),
					},
					H2: {
						value: Text,
						styles: (theme: ThemeValue) => ({ root: theme.typography.h2 }),
					},
					H3: {
						value: Text,
						styles: (theme: ThemeValue) => ({ root: theme.typography.h3 }),
					},
					H4: {
						value: Text,
						styles: (theme: ThemeValue) => ({ root: theme.typography.h4 }),
					},
					H5: {
						value: Text,
						styles: (theme: ThemeValue) => ({ root: theme.typography.h5 }),
					},
					H6: {
						value: Text,
						styles: (theme: ThemeValue) => ({ root: theme.typography.h6 }),
					},
					Subtitle1: {
						value: Text,
						styles: (theme: ThemeValue) => ({ root: theme.typography.subtitle1 }),
					},
					Subtitle2: {
						value: Text,
						styles: (theme: ThemeValue) => ({ root: theme.typography.subtitle2 }),
					},
					Body1: {
						value: Text,
						styles: (theme: ThemeValue) => ({ root: theme.typography.body1 }),
					},
					Body2: {
						value: Text,
						styles: (theme: ThemeValue) => ({ root: theme.typography.body2 }),
					},
					Caption: {
						value: Text,
						styles: (theme: ThemeValue) => ({ root: theme.typography.caption }),
					},
					Overline: {
						value: Text,
						styles: (theme: ThemeValue) => ({ root: theme.typography.overline }),
					},
				});

				return bootOptions;
			},
		},
	],

	/**
	 * This filter registers components from bootOptions.components property.
	 * These are the components typically set in the bluebase.js file.
	 */
	'bluebase.components.register': [
		{
			key: 'bluebase-components-register-default',
			priority: 5,

			// tslint:disable-next-line:object-literal-sort-keys
			value: async (bootOptions: BootOptions, _ctx: {}, BB: BlueBase) => {
				await BB.Components.registerCollection(bootOptions.components);

				return bootOptions;
			},
		},
	],
};
