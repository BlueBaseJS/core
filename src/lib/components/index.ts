import {
	ErrorStateProps,
	LoadingStateProps,
	StatefulComponentProps
} from '../../components';
import { TextProps, ViewProps } from 'react-native';
import { BlueBase } from '../../BlueBase';
import { BlueBaseContext } from '../../Context';
import { ButtonProps } from '../../ui-interfaces';
import React from 'react';

export function getComponent<T = any>(name: string) {
	return class BlueBaseComponent extends React.PureComponent<T> {

		static contextType = BlueBaseContext;

		render() {
			const BB: BlueBase = this.context;
			const Component = BB.Components.resolve(name);

			return React.createElement(Component, this.props);
		}
	} as React.ComponentType<T>;
}

// BlueBase
export const ErrorState = getComponent<ErrorStateProps>('ErrorState');
export const LoadingState = getComponent<LoadingStateProps>('LoadingState');
export const StatefulComponent = getComponent<StatefulComponentProps>('StatefulComponent');

// Native
export const Button = getComponent<ButtonProps>('Button');
export const Text = getComponent<TextProps>('Text');
export const View = getComponent<ViewProps>('View');

