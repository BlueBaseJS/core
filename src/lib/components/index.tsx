import { ActivityIndicatorProps, ImageProps, TextProps, ViewProps } from '../../native';
import { BlueBaseConsumer } from '../../Context';
import {
	BlueBaseHookProps,
	ComponentStateProps,
	DataObserverProps,
	DynamicIconProps,
	ErrorObserverProps,
	ErrorStateProps,
	HoverObserverProps,
	JsonSchemaProps,
	LoadingStateProps,
	PluginIconProps,
	StatefulComponentProps,
	WaitObserverProps,
} from '../../components';
import { BlueBase } from '../../BlueBase';
import { ButtonProps } from '../../ui-interfaces';
import React from 'react';

export function getComponent<T = any>(name: string) {

	const BlueBaseComponent = (props: T) => (
		<BlueBaseConsumer children={(BB: BlueBase) => {
			const Component = BB.Components.resolve(name);

			return React.createElement(Component, props);
		}} />
	);

	BlueBaseComponent.displayName = name;

	return BlueBaseComponent as React.ComponentType<T>;
}

// BlueBase
export const BlueBaseHook = getComponent<BlueBaseHookProps>('BlueBaseHook');
export const ComponentState = getComponent<ComponentStateProps>('ComponentState');
export const DataObserver = getComponent<DataObserverProps>('DataObserver');
export const DynamicIcon = getComponent<DynamicIconProps>('DynamicIcon');
export const EmptyState = getComponent<{}>('EmptyState');
export const ErrorObserver = getComponent<ErrorObserverProps>('ErrorObserver');
export const ErrorState = getComponent<ErrorStateProps>('ErrorState');
export const HoverObserver = getComponent<HoverObserverProps>('HoverObserver');
export const JsonSchema = getComponent<JsonSchemaProps>('JsonSchema');
export const LoadingState = getComponent<LoadingStateProps>('LoadingState');
export const PluginIcon = getComponent<PluginIconProps>('PluginIcon');
export const StatefulComponent = getComponent<StatefulComponentProps>('StatefulComponent');
export const WaitObserver = getComponent<WaitObserverProps>('WaitObserver');

// Native
export const ActivityIndicator = getComponent<ActivityIndicatorProps>('ActivityIndicator');
export const Button = getComponent<ButtonProps>('Button');
export const Image = getComponent<ImageProps>('Image');
export const Text = getComponent<TextProps>('Text');
export const View = getComponent<ViewProps>('View');

// Typography
export const H1 = getComponent<TextProps>('H1');
export const H2 = getComponent<TextProps>('H2');
export const H3 = getComponent<TextProps>('H3');
export const H4 = getComponent<TextProps>('H4');
export const H5 = getComponent<TextProps>('H5');
export const H6 = getComponent<TextProps>('H6');
export const Subtitle1 = getComponent<TextProps>('Subtitle1');
export const Subtitle2 = getComponent<TextProps>('Subtitle2');
export const Body1 = getComponent<TextProps>('Body1');
export const Body2 = getComponent<TextProps>('Body2');
export const Caption = getComponent<TextProps>('Caption');
export const Overline = getComponent<TextProps>('Overline');
