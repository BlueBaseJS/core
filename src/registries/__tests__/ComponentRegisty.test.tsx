import { Text, View } from 'react-native';
import { BlueBase } from '../../BlueBase';
import { BlueBaseApp } from '../../components';
import { ComponentRegistry } from '../';
import React from 'react';
import TestRenderer from 'react-test-renderer';
import { getComponent } from '../../getComponent';
import { mount } from 'enzyme';

const Button: React.StatelessComponent<{}> = () => <View>A Button</View>;
(Button as any).foo = 'bar';

class Welcome extends React.Component<{ name: string }> {
	render() {
		return <h1>Hello, {this.props.name}</h1>;
	}
}

describe('ComponentRegistry', () => {


	describe('Proxy', () => {

		it('should resolve a component through proxy', async () => {
			const BB = new BlueBase();
			const Components = new ComponentRegistry(BB);
			await Components.register('Button', Button);

			expect(getComponent('Button')).toBeTruthy();
		});

		it('should throw for unknown components', async () => {
			const BB = new BlueBase();
			const Components = new ComponentRegistry(BB);
			expect(() => Components.resolve('Foo')).toThrow();
		});

	});

	describe('.resolve method', () => {

		it('should throw for unknown components', async () => {
			const BB = new BlueBase();
			const Components = new ComponentRegistry(BB);

			let message = false;

			try {
				Components.resolve('Foo');
			} catch (e) {
				message = e.message;
			}

			expect(message).toBe('Could not resolve any of the following components: [Foo].');
		});

		it('should resolve a component', async () => {
			const BB = new BlueBase();
			const Components = new ComponentRegistry(BB);

			await Components.register('Foo', Text);
			const Foo = Components.resolve('Foo');

			const rendered = TestRenderer.create(<Foo>Some text</Foo>);

			expect(rendered.root.props.children).toBe('Some text');
			// expect(rendered.root.type.displayName).toBe('Text');
			expect(rendered.toJSON()).toMatchSnapshot();
		});

		it('should resolve a Promised component', async (done) => {
			const Foo = getComponent('Foo');
			const wrapper = mount(
				<BlueBaseApp components={{ Foo: Promise.resolve(Text) as any }}>
					<Foo>Some text</Foo>
				</BlueBaseApp>
			);

			// expect(json).toBe(null);
			expect(wrapper).toMatchSnapshot();

			setTimeout(() => {
				wrapper.update();
				expect(wrapper).toMatchSnapshot();

				// Verify that background color is light
				const view = wrapper.find('Foo').find('Text').last();
				expect(view.text()).toBe('Some text');

				done();
			});
		});

		it('should add an HOC that makes background green', async () => {
			const BB = new BlueBase();
			const Components = new ComponentRegistry(BB);

			await Components.register('Button', Button);
			Components.addHocs('Button', (Comp: React.ComponentType) => () => (
				<View style={{ backgroundColor: 'green' }}>
					<Comp />
				</View>
			));

			const ResolvedComponent = Components.resolve('Button');

			const rendered = TestRenderer.create(<ResolvedComponent />);
			const tree = rendered.toJSON() as any;
			expect(tree.props.style.backgroundColor).toBe('green');
			expect(tree).toMatchSnapshot();
		});

		it('should add an HOC with its arguments that makes background orange', async () => {
			const BB = new BlueBase();
			const Components = new ComponentRegistry(BB);

			await Components.register('Button', Button);

			const hoc = (args: { backgroundColor: string }) =>
			(Comp: React.ComponentType) =>
			() => (
				<View style={{ backgroundColor: args.backgroundColor }}>
					<Comp />
				</View>
			);

			Components.addHocs('Button', [hoc, { backgroundColor: 'orange' }]);

			const ResolvedComponent = Components.resolve('Button');

			const rendered = TestRenderer.create(<ResolvedComponent />);
			const tree = rendered.toJSON() as any;
			expect(tree.props.style.backgroundColor).toBe('orange');
			expect(tree).toMatchSnapshot();
		});

		it('should not apply styles if applyStyles flag is set false', async () => {
			const BB = new BlueBase();
			const Components = new ComponentRegistry(BB);

			await Components.register({
				applyStyles: false,
				key: 'Button',
				value: Button,
			});

			const ResolvedComponent = Components.resolve('Button');

			const rendered = TestRenderer.create(<ResolvedComponent />);
			const tree = rendered.toJSON() as any;
			expect(tree.props.styles).toBeUndefined();
			expect(tree).toMatchSnapshot();
		});

		it('should hoist static properties of the original component', async () => {
			const BB = new BlueBase();
			const Components = new ComponentRegistry(BB);

			await Components.register('Button', Button);

			const hoc = (args: { backgroundColor: string }) =>
			(Comp: React.ComponentType) =>
			() => (
				<View style={{ backgroundColor: args.backgroundColor }}>
					<Comp />
				</View>
			);

			Components.addHocs('Button', [hoc, { backgroundColor: 'orange' }]);

			const ResolvedComponent = Components.resolve('Button');

			expect((ResolvedComponent as any).foo).toBe('bar');
		});

	});

	describe('.register method', () => {

		/////////////////////////////////
		//////// React Component ////////
		/////////////////////////////////

		it('should register a React Functional Component', async () => {
			const BB = new BlueBase();
			const Components = new ComponentRegistry(BB);
			await Components.register('Button', Button);

			// TODO: Add proper dom testing
			expect(getComponent('Button')).toBeTruthy();
		});

		it('should register a React Functional Component in an ES Module', async () => {
			const BB = new BlueBase();
			const Components = new ComponentRegistry(BB);
			await Components.register('Button', { default: Button } as any);

			// TODO: Add proper dom testing
			expect(getComponent('Button')).toBeTruthy();
		});

		it('should register a React Functional Component in an Promised ES Module', async () => {
			const BB = new BlueBase();
			const Components = new ComponentRegistry(BB);

			await Components.register('Button', Promise.resolve({ default: Button }) as any);

			// TODO: Add proper dom testing
			expect(getComponent('Button')).toBeTruthy();
		});

		it('should register a React Class Component', async () => {
			const BB = new BlueBase();
			const Components = new ComponentRegistry(BB);

			await Components.register('Welcome', Welcome);

			// TODO: Add proper dom testing
			expect(getComponent('Welcome')).toBeTruthy();
		});

		it('should register a React Class Component in an ES Module', async () => {
			const BB = new BlueBase();
			const Components = new ComponentRegistry(BB);

			await Components.register('Welcome', { default: Welcome } as any);

			// TODO: Add proper dom testing
			expect(getComponent('Welcome')).toBeTruthy();
		});

		it('should register a React Class Component in an Promised ES Module', async () => {
			const BB = new BlueBase();
			const Components = new ComponentRegistry(BB);

			await Components.register('Welcome', Promise.resolve({ default: Welcome }) as any);

			// TODO: Add proper dom testing
			expect(getComponent('Welcome')).toBeTruthy();
		});

		///////////////////////////////////////
		//////// ComponentRegistryItem ////////
		///////////////////////////////////////

		it('should register a ComponentRegistryItem', async () => {
			const BB = new BlueBase();
			const Components = new ComponentRegistry(BB);

			await Components.register('Button', {
				value: Button
			});

			// TODO: Add proper dom testing
			expect(getComponent('Button')).toBeTruthy();
		});

		it('should register a ComponentRegistryItem, with value in an ES Module', async () => {
			const BB = new BlueBase();
			const Components = new ComponentRegistry(BB);

			await Components.register('Button', {
				value: { default: Button } as any
			});

			// TODO: Add proper dom testing
			expect(getComponent('Button')).toBeTruthy();
		});


		it('should register a ComponentRegistryItem, with value in a Promised ES Module', async () => {
			const BB = new BlueBase();
			const Components = new ComponentRegistry(BB);

			await Components.register('Button', {
				value: Promise.resolve({ default: Button }) as any
			});

			// TODO: Add proper dom testing
			expect(getComponent('Button')).toBeTruthy();
		});

		it('should register a ComponentRegistryItem, with preload set to true', async () => {
			const BB = new BlueBase();
			const Components = new ComponentRegistry(BB);

			await Components.register('Button', {
				preload: true,
				value: Button,
			});

			const Component = Components.get('Button');

			if (!Component) {
				throw Error();
			}

			expect(Component.preload).toBe(true);
		});

		it('should throw an Error for empty object', async () => {
			const BB = new BlueBase();
			const Components = new ComponentRegistry(BB);


			let message = false;

			try {
				await Components.register('Button', {} as any);
			} catch (e) {
				message = e.message;
			}

			expect(message).toBe('Could not register item. Reason: Unknown item type.');
		});


		////// Overrides

		it('should merge a ComponentRegistryItem with an existing one', async () => {
			const BB = new BlueBase();
			const Components = new ComponentRegistry(BB);

			await Components.register('Button', {
				hocs: [() => Button],
				preload: true,
				value: Button,
			});
			await Components.register('Button', {
				hocs: [() => Button, () => Button],
				value: Button,
			});

			const Component = Components.get('Button');

			if (!Component) {
				throw Error();
			}
			expect(Component.hocs).toHaveLength(3);
			expect(Component.preload).toBe(true);
		});
	});


	describe('.addHocs method', () => {

		it('should add a single HOC', async () => {
			const BB = new BlueBase();
			const Components = new ComponentRegistry(BB);

			await Components.register('Button', Button);
			Components.addHocs('Button', () => Button);

			const Component = Components.get('Button');

			if (!Component) {
				throw Error();
			}
			expect(Component.hocs).toHaveLength(1);
		});

		it('should add multiple HOCs', async () => {
			const BB = new BlueBase();
			const Components = new ComponentRegistry(BB);

			await Components.register('Button', Button);
			Components.addHocs('Button', () => Button);
			Components.addHocs('Button', () => Button);
			Components.addHocs('Button', () => Button);

			const Component = Components.get('Button');

			if (!Component) {
				throw Error();
			}
			expect(Component.hocs).toHaveLength(3);
		});

		it('should not override existing HOCs', async () => {
			const BB = new BlueBase();
			const Components = new ComponentRegistry(BB);

			await Components.register('Button', {
				hocs: [() => Button],
				value: Button,
			});
			Components.addHocs('Button', () => Button);

			const Component = Components.get('Button');

			if (!Component) {
				throw Error();
			}
			expect(Component.hocs).toHaveLength(2);
		});

		it('should throw an Error for unregistered components', async () => {
			const BB = new BlueBase();
			const Components = new ComponentRegistry(BB);


			let message = false;

			try {
				Components.addHocs('Button', () => Button);
			} catch (e) {
				message = e.message;
			}

			expect(message).toBe('Could not add hocs for "Button" component. Reason: Component not found.');
		});

	});


	describe('.setStyles/getStyles method', () => {

		it('should set & get styles of a Component', async () => {
			const BB = new BlueBase();
			const Components = new ComponentRegistry(BB);

			await Components.register('Button', Button);
			Components.setStyles('Button', {
				root: {
					backgroundColor: 'blue',
				}
			});

			const styles = Components.getStyles('Button');

			if (!styles) {
				throw Error();
			}
			expect(styles).toMatchObject({
				root: {
					backgroundColor: 'blue',
				}
			});
		});

		it('should throw an Error if setting styles of an unknown component', async () => {
			const BB = new BlueBase();
			const Components = new ComponentRegistry(BB);

			try {
				Components.setStyles('Button', {
					root: {
						backgroundColor: 'blue',
					}
				});
			} catch (error) {
				expect(error.message).toBe('Cannot set styles "Button" component. Reason: Component not found.');
			}

		});

		it('should return undefined if getting styles of an unknown component', async () => {
			const BB = new BlueBase();
			const Components = new ComponentRegistry(BB);
			expect(Components.getStyles('Button')).toBe(undefined);
		});

	});

});