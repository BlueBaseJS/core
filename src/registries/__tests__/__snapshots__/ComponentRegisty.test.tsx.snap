// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComponentRegistry .resolve method should add an HOC that makes background green 1`] = `
<View
  style={
    Object {
      "backgroundColor": "green",
    }
  }
/>
`;

exports[`ComponentRegistry .resolve method should add an HOC with its arguments that makes background orange 1`] = `
<View
  style={
    Object {
      "backgroundColor": "orange",
    }
  }
/>
`;

exports[`ComponentRegistry .resolve method should resolve a Promised component 1`] = `
<BlueBaseApp
  components={
    Object {
      "Foo": Promise {
        "_40": 0,
        "_55": [Function],
        "_65": 1,
        "_72": null,
      },
    }
  }
>
  <Component>
    <Text>
      Loading
    </Text>
  </Component>
</BlueBaseApp>
`;

exports[`ComponentRegistry .resolve method should resolve a Promised component 2`] = `
<BlueBaseApp
  components={
    Object {
      "Foo": Promise {
        "_40": 0,
        "_55": [Function],
        "_65": 1,
        "_72": null,
      },
    }
  }
>
  <BlueBaseRoot>
    <ThemeProvider>
      <SystemApp>
        <ThemedComponent>
          <SystemApp>
            <Foo>
              <ThemedComponent>
                <Component>
                  <Text>
                    Some text
                  </Text>
                </Component>
              </ThemedComponent>
            </Foo>
          </SystemApp>
        </ThemedComponent>
      </SystemApp>
    </ThemeProvider>
  </BlueBaseRoot>
</BlueBaseApp>
`;

exports[`ComponentRegistry .resolve method should resolve a component 1`] = `null`;
